src/vos/vos_obj.c

```diff
@@ -51,10 +102,43 @@ static int
 tree_is_empty(struct vos_object *obj, umem_off_t *known_key, daos_handle_t toh,
 	      const daos_epoch_range_t *epr, vos_iter_type_t type)
 {
+	daos_anchor_t		 anchor = {0};
 	struct dtx_handle	*dth = vos_dth_get();
+	d_iov_t			 key;
 	struct vos_key_info	 kinfo = {0};
+	struct vos_krec_df	*krec;
 	int			 rc;
 
+	/** The address of the known_key, which actually points at the krec is guaranteed by PMDK
+	 *  to be allocated at an 8 byte alignment so the low order bit is available to mark it as
+	 *  punched.
+	 */
+	if (*known_key != UMOFF_NULL && (*known_key & 0x1) == 0)
+		return 0;
+
+	kinfo.ki_obj = obj;
+	kinfo.ki_known_key = known_key;
+
+	if (*known_key == UMOFF_NULL)
+		goto tail;
+
+	krec = umem_off2ptr(vos_obj2umm(obj), (*known_key & ~(1ULL)));
+	d_iov_set(&key, vos_krec2key(krec), krec->kr_size);
+	dbtree_key2anchor(toh, &key, &anchor);
+
+	rc = vos_iterate_key(obj, toh, type, epr, true, empty_tree_check,
+			     &kinfo, dth, &anchor);
+
+	if (rc < 0)
+		return rc;
+
+	if (kinfo.ki_non_empty)
+		return 0;
+
+	/** Start from beginning one more time.  It will iterate until it
+	 *  sees the first thing it saw
```

Comment:
What is the benefit to iterate again from the beginning? Why not directly goto the beginning when do the first iteration? We do not care which entry will be found, we only care whether there is something in the tree. Then starting the iteration from which point is not important, right? If yes, we can avoid double scan?

Reply:
A common case is removal of all keys.  If we start from a known removed key, we don't iterate the million keys that have already been deleted on every punch.  We only do it once on the final deletion.

It ultimately turns an O(n^2) punch of all keys into O(n) because during the punch phase, we will only visit each key once.